lib/uri_utils.ml

```diff
@@ -1,9 +1,42 @@
 open Import
 
-let has_git_extension uri =
+let path_ok path = path |> Fpath.to_string |> Result.ok
+let flip f a b = f b a
+
+let dump_result =
+  let dump_msg ppf (`Msg s) = Fmt.pf ppf "`Msg %S" s in
+  Fmt.Dump.result ~ok:Fmt.Dump.string ~error:dump_msg
+
+let canonicalize uri =
   let open Result.O in
-  let ext_res =
-    let+ path = Fpath.of_string (Uri.path uri) in
-    Fpath.get_ext ~multi:true path
+  let new_path =
+    let* fpath = Fpath.of_string (Uri.path uri) in
+    let fpath =
+      match Fpath.has_ext ".git" fpath with
+      | true -> fpath
+      | false -> Fpath.add_ext ".git" fpath
```

Comment:
Is it possible that the uri refers to a non-git repo?

Reply:
In theory it can, but

1. Most `dev-repo`s these days are git and even on github.
2. It doesn't matter all that much because this is only to make the URLs canonical for comparison, so `https://hghub.com/foo/bar.hg` would be rewritten to `https+git://hghub.com/foo/bar.hg.git` which might not be a valid URL but will compare equal to itself.

The canonicalization process adds `.git` and rewrites schemes to `git+https` but if you'd rather have it remove `.git` and rewrite `git+https://` and `git://` to `https://` that's also totally ok with me.